onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,563 @@
+                                 Provider_OpKernelInfo_Impl provider_info(info);
+                                 return new OpKernel_Translator(provider_info, kernel_create_func(provider_info));
+                               });
+
+    return p_->Register(std::move(info_real));
+  }
+
+  std::shared_ptr<KernelRegistry> p_owned_;
+  KernelRegistry* p_{&*p_owned_};
+};
+
+struct Provider_IExecutionProvider_Router_Impl : Provider_IExecutionProvider_Router, IExecutionProvider {
+  Provider_IExecutionProvider_Router_Impl(Provider_IExecutionProvider* outer, const std::string& type) : IExecutionProvider(type), outer_(outer) {
+  }
+
+  virtual ~Provider_IExecutionProvider_Router_Impl() {}
+
+  std::shared_ptr<Provider_KernelRegistry> Provider_GetKernelRegistry() const override {
+    return std::make_shared<Provider_KernelRegistry_Impl>(GetKernelRegistry());
+  }
+
+  std::shared_ptr<KernelRegistry> GetKernelRegistry() const override {
+    return static_cast<Provider_KernelRegistry_Impl*>(&*outer_->Provider_GetKernelRegistry())->p_owned_;
+  }
+
+  std::vector<std::unique_ptr<Provider_ComputeCapability>> Provider_GetCapability(const onnxruntime::Provider_GraphViewer& graph,
+                                                                                  const std::vector<const Provider_KernelRegistry*>& kernel_registries) const override {
+    std::vector<const KernelRegistry*> kernel_registries_internal;
+    for (auto& v : kernel_registries)
+      kernel_registries_internal.emplace_back(static_cast<const Provider_KernelRegistry_Impl*>(v)->p_);
+
+    auto capabilities_internal = IExecutionProvider::GetCapability(static_cast<const Provider_GraphViewer_Impl*>(&graph)->v_, kernel_registries_internal);
+
+    std::vector<std::unique_ptr<Provider_ComputeCapability>> capabilities;
+    for (auto& v : capabilities_internal)
+      capabilities.emplace_back(onnxruntime::make_unique<Provider_ComputeCapability>(onnxruntime::make_unique<Provider_IndexedSubGraph_Impl>(std::move(v->sub_graph))));
+    return capabilities;
+  }
+
+  std::vector<std::unique_ptr<ComputeCapability>> GetCapability(const onnxruntime::GraphViewer& graph,
+                                                                const std::vector<const KernelRegistry*>& kernel_registries) const override {
+    std::vector<const Provider_KernelRegistry*> registries;
+    for (auto p : kernel_registries)
+      registries.push_back(new Provider_KernelRegistry_Impl(const_cast<KernelRegistry*>(p)));
+
+    auto provider_result = outer_->Provider_GetCapability(Provider_GraphViewer_Impl(graph), registries);
+    std::vector<std::unique_ptr<ComputeCapability>> result;
+
+    for (auto& p : provider_result)
+      result.emplace_back(onnxruntime::make_unique<ComputeCapability>(std::move(static_cast<Provider_IndexedSubGraph_Impl*>(p->t_sub_graph_.get())->p_)));
+
+    for (auto p : registries)
+      delete p;
```

Comment:
nit: I assume using RAII was not easy here?

Reply:
Yeah, the interface to the provider was just an array of pointers. I'd have to change both sidesto use unique_ptr vectors. It's ugly right now, but if we make the interfaces go deeper than this layer this problem will go away.